### PR TITLE
Feat/marketplace precompile trn 371

### PR DESCRIPTION
**Pull request by KarishmaBothara**
**PR checklist**  
Marketplace precompile SFT support

### CODE REVIEW

File: evm-precompiles/marketplace/README.md [hunk 1]
```diff
@@ -7,34 +7,38 @@ Precompile address: `0x00000000000000000000000000000000000006CD`
...
+    event FixedPriceSaleClose(uint256 indexed collectionId, uint256 indexed listingId, address sender, uint256[] serialNumbers, uint128 marketplace_id); //uint256;uint256;address;uint256[]
+    event AuctionClose(uint256 indexed collectionId, uint256 indexed listingId, address sender, uint256[] serialNumbers, uint128 marketplace_id);
+    event Offer(uint256 indexed offerId, address indexed sender, uint256 collectionId, uint256 seriesId, uint128 marketplace_id);
+    event OfferCancel(uint256 indexed offerId, address indexed sender, uint256 collectionId, uint256 seriesId, uint128 marketplace_id);
+    event OfferAccept(uint256 indexed offerId, uint256 indexed amount, address indexed sender, uint256 collectionId, uint256 seriesId, uint128 marketplace_id); // uint256;uint256;address;uint256
+
     function registerMarketplace(address marketplaceAccount, uint256 entitlement) external returns (uint marketplaceId);
     function sellNftWithMarketplaceId(address collectionAddress, uint256[] calldata serialNumberIds, address buyer, address paymentAsset, uint256 fixedPrice, uint256 duration, uint32 marketplaceId) external returns (uint listingId);
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/README.md:22]: Please can you add a deprecated flag on these old functions?

File: evm-precompiles/marketplace/src/lib.rs [hunk 2]
```diff
@@ -23,48 +23,69 @@ use frame_support::{
...
+use sp_std::{marker::PhantomData, vec::Vec};
 
 /// Solidity selector of the Marketplace register log, which is the Keccak of the Log signature.
 pub const SELECTOR_LOG_MARKETPLACE_REGISTER: [u8; 32] =
 	keccak256!("MarketplaceRegister(address,uint256,address)"); // caller_id, marketplace_id
 
-pub const SELECTOR_LOG_FIXED_PRICE_SALE_LIST: [u8; 32] =
-	keccak256!("FixedPriceSaleList(address,uint256,uint256,uint256[],address)"); // seller_id, listing_id, fixed_price, serial_number_ids, collection_address
+pub const SELECTOR_LOG_FIXED_PRICE_SALE_LIST_NFT: [u8; 32] =
+	keccak256!("FixedPriceSaleListForNFT(address,uint256,uint256,uint256[],address,uint128)");
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:46]: nit. rename to `FixedPriceSaleListNFT`

File: evm-precompiles/marketplace/src/lib.rs [hunk 3]
```diff
@@ -23,48 +23,69 @@ use frame_support::{
...
+// seller_id, listing_id, fixed_price, serial_number_ids, collection_address, marketplace_id
+
+pub const SELECTOR_LOG_FIXED_PRICE_SALE_LIST_SFT: [u8; 32] =
+	keccak256!("FixedPriceSaleListForSFT(address,uint256,uint256,uint256[],address,uint128)");
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:50]: nit. Rename to FixedPriceSaleListSFT
Also we should add a field for the amounts that were listed (*i.e. (1124, 5) x 100)

File: evm-precompiles/marketplace/src/lib.rs [hunk 4]
```diff
@@ -23,48 +23,69 @@ use frame_support::{
...
+	keccak256!("FixedPriceSaleUpdate(uint256,uint256,uint256,address,uint256[],uint128)");
+// collection_id, listing_id, new_price, sender, serial_number_ids, marketplace_id
 
 pub const SELECTOR_LOG_FIXED_PRICE_SALE_COMPLETE: [u8; 32] =
-	keccak256!("FixedPriceSaleComplete(uint256,uint256,uint256,address,uint256[])"); // collection_id, listing_id, fixed_price, sender, serial_number_ids
+	keccak256!("FixedPriceSaleComplete(uint256,uint256,uint256,address,uint256[],uint128)");
+// collection_id, listing_id, fixed_price, sender, serial_number_ids, marketplace_id
+
+pub const SELECTOR_LOG_AUCTION_OPEN_NFT: [u8; 32] =
+	keccak256!("AuctionOpenForNFT(uint256,uint256,uint256,address,uint256[],uint128)");
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:62]:
```suggestion
	keccak256!("AuctionOpenNFT(uint256,uint256,uint256,address,uint256[],uint128)");
```

File: evm-precompiles/marketplace/src/lib.rs [hunk 5]
```diff
@@ -23,48 +23,69 @@ use frame_support::{
...
+// collection_id, listing_id, reserve_price, sender, serial_number_ids, marketplace_id
 
-pub const SELECTOR_LOG_AUCTION_OPEN: [u8; 32] =
-	keccak256!("AuctionOpen(uint256,uint256,uint256,address,uint256[])"); // collection_id, listing_id, reserve_price, sender, serial_number_ids
+pub const SELECTOR_LOG_AUCTION_OPEN_SFT: [u8; 32] =
+	keccak256!("AuctionOpenForSFT(uint256,uint256,uint256,address,uint256[],uint128)");
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:66]:
```suggestion
	keccak256!("AuctionOpenSFT(uint256,uint256,uint256,address,uint256[],uint128)");
```

File: evm-precompiles/marketplace/src/lib.rs [hunk 6]
```diff
@@ -287,37 +318,133 @@ where
...
 			buyer,
 			payment_asset,
 			fixed_price,
 			duration,
+			Some(marketplace_id),
+			Some(quantities),
+		)
+	}
+
+	fn sell_internal_sft(
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:355]: Now that we no longer have the two functions, it doesn't make as much sense to separate out the internal function. All logic should be within sell_sft and sell_nft etc

File: evm-precompiles/marketplace/src/lib.rs [hunk 7]
```diff
@@ -287,37 +318,133 @@ where
...
+		handle: &mut impl PrecompileHandle,
+		collection_address: Address,
+		serial_number_ids: Vec<U256>,
+		buyer: Address,
+		payment_asset: Address,
+		fixed_price: U256,
+		duration: U256,
+		marketplace_id: Option<MarketplaceId>,
+		quantities: Option<Vec<U256>>,
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:364]: Why make quantities an option?

File: evm-precompiles/marketplace/src/lib.rs [hunk 8]
```diff
@@ -287,37 +318,133 @@ where
...
+			<Runtime as ErcIdConversion<CollectionUuid>>::evm_id_to_runtime_id(
+				collection_address,
+				ERC1155_PRECOMPILE_ADDRESS_PREFIX,
+			)
+			.ok_or_else(|| revert("Marketplace SFT: Invalid collection address"))?;
+
+		let serials_unbounded = serial_number_ids
+			.clone()
+			.into_iter()
+			.zip(quantities.unwrap())
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:389]: Need to assert the length of serial_numbers and quantities are equal and greater than 0 prior to zipping

File: evm-precompiles/marketplace/src/lib.rs [hunk 9]
```diff
@@ -287,37 +318,133 @@ where
...
+					return Err(revert("Marketplace SFT: Expected quantity <= 2^128").into())
+				}
+				let serial_number: SerialNumber = serial_number.saturated_into();
+				let quantity: Balance = quantity.saturated_into();
+				Ok((serial_number, quantity))
+			})
+			.collect::<Result<Vec<(SerialNumber, Balance)>, PrecompileFailure>>()?;
+
+		let serial_numbers: BoundedVec<(SerialNumber, Balance), Runtime::MaxTokensPerListing> =
+			BoundedVec::truncate_from(serials_unbounded);
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:404]: We can't use a truncate_from here as the length of the serial numbers is not checked against the upper bound. try_from should be used instead

File: evm-precompiles/marketplace/src/lib.rs [hunk 10]
```diff
@@ -287,37 +318,133 @@ where
...
+		let tokens = ListingTokens::Sft(SftListing { collection_id, serial_numbers });
+		let buyer: H160 = buyer.into();
+		let buyer: Option<Runtime::AccountId> =
+			if buyer == H160::default() { None } else { Some(buyer.into()) };
+
+		let caller: Runtime::AccountId = handle.context().caller.into();
+		// Manually record gas
+		handle.record_cost(Runtime::GasWeightMapping::weight_to_gas(
+			<Runtime as pallet_marketplace::Config>::WeightInfo::sell_nft(
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:413]:
```suggestion
			<Runtime as pallet_marketplace::Config>::WeightInfo::sell_sft(
```

File: evm-precompiles/marketplace/src/lib.rs [hunk 11]
```diff
@@ -632,13 +784,105 @@ where
...
+
+		// Parse asset_id
+		let payment_asset: AssetId = <Runtime as ErcIdConversion<AssetId>>::evm_id_to_runtime_id(
+			payment_asset,
+			ERC20_PRECOMPILE_ADDRESS_PREFIX,
+		)
+		.ok_or_else(|| revert("Marketplace SFT: Invalid payment asset address"))?;
+
+		handle.record_cost(Runtime::GasWeightMapping::weight_to_gas(
+			<Runtime as pallet_marketplace::Config>::WeightInfo::auction_nft(
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:858]:
```suggestion
			<Runtime as pallet_marketplace::Config>::WeightInfo::auction_sft(
```

File: evm-precompiles/marketplace/src/lib.rs [hunk 12]
```diff
@@ -632,13 +784,105 @@ where
...
+					return Err(revert("Marketplace: Expected quantity <= 2^128").into())
+				}
+				let serial_number: SerialNumber = serial_number.saturated_into();
+				let quantity: Balance = quantity.saturated_into();
+				Ok((serial_number, quantity))
+			})
+			.collect::<Result<Vec<(SerialNumber, Balance)>, PrecompileFailure>>()?;
+
+		let serial_numbers: BoundedVec<(SerialNumber, Balance), Runtime::MaxTokensPerListing> =
+			BoundedVec::truncate_from(serials_unbounded);
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:845]: Same as before, should be try_from

File: evm-precompiles/marketplace/src/lib.rs [hunk 13]
```diff
@@ -632,13 +784,105 @@ where
...
+			<Runtime as ErcIdConversion<CollectionUuid>>::evm_id_to_runtime_id(
+				collection_address,
+				ERC1155_PRECOMPILE_ADDRESS_PREFIX,
+			)
+			.ok_or_else(|| revert("Marketplace: Invalid collection address"))?;
+
+		let serials_unbounded = serial_number_ids
+			.clone()
+			.into_iter()
+			.zip(quantities.unwrap())
```
Comment (JasonTulp) [inline evm-precompiles/marketplace/src/lib.rs:830]: Need to check that quantities.len() == serial_numbers.len()

File: evm-precompiles/marketplace/src/lib.rs (hunk 3 above)
Comment (KarishmaBothara) [reply evm-precompiles/marketplace/src/lib.rs:50]: amounts as in quantities?

File: evm-precompiles/marketplace/src/lib.rs (hunk 9 above)
Comment (KarishmaBothara) [reply evm-precompiles/marketplace/src/lib.rs:404]: try_from doesn't work... I used truncate_from cause I saw something similar here - https://github.com/futureversecom/trn-seed/blob/d241c0cc1753ac8c52d44f264d0fc9d262d96ae7/pallet/sft/src/tests.rs#L83

**Review by JasonTulp — APPROVED** [review]
LGTM!

**Pull request merged**